Mathlib/Algebra/Polynomial/RuleOfSigns.lean

```diff
@@ -0,0 +1,588 @@
+          eraseLead_mul_eq_mul_eraseLead_of_nextCoeff_zero hη.ne' <|
+            sign_eq_zero_iff.mp (SignType.zero_eq_zero ▸ hcsq1 ▸ hsq1)
+        have h_e2LQ := signvar_eraseLead_le (eraseLead ((X - C η) * P))
+        rw [h_elMul] at h_e2LQ
+        linarith [signvar_le_eraseLead_succ P, ih _ hnDeQ h_eQ_nz rfl]
+    all_goals (
+      have h₁ : nextCoeff P ≠ 0 := by
+        simp [← sign_ne_zero, ← hsq1, hcsq1]
+
+      specialize ih _ hnDeQ (mt nextCoeff_eq_zero_of_eraseLead_eq_zero h₁) rfl
+
+      have h_eQ : P.SignVariations = P.eraseLead.SignVariations + ?_ := by
+        simp [signvar_eq_eraseLead_add_ite hq, leadingCoeff_eraseLead_eq_nextCoeff h₁,
+          ← hsq0, ← hsq1, h_sq0_pos, hcsq1]
+        exact rfl
+    )
+    · --sq1 is negative
+      -- P starts with [+,-,...], so (X-C)*P starts with [+,-,...].
+      -- After dropping the lead of P, this becomes [-,...] and [-,...].
+      -- Dropping the first one of each decreases (X-C)*P by one and P by one, so we can induct.
+      have hηq1 : sηq1 = -1 := by
+        apply sign_eq_neg_one_iff.mpr
+        have hc : coeff P d < 0 := by
+          rw [(Nat.sub_eq_of_eq_add hd.symm).symm]
+          simp [nextCoeff, hcsq1, hd ▸ hd0] at hsq1
+          exact sign_eq_neg_one_iff.mp hsq1.symm
+        rw [nextCoeff, natDegree_mul (X_sub_C_ne_zero η) hq, natDegree_X_sub_C]
+        simp only [add_eq_zero, one_ne_zero, false_and, add_tsub_cancel_left, ite_false]
+        rw [← hd, coeff_X_sub_C_mul, sub_neg]
+        exact hc.trans hηc
+
+      have h₂ : ((X - C η) * P).eraseLead.leadingCoeff = ((X - C η) * P).nextCoeff := by
+        refine leadingCoeff_eraseLead_eq_nextCoeff (fun h ↦ ?_)
+        simp [h, hsηq1] at hηq1
+
+      have hLQ : ((X - C η) * P).SignVariations = ((X - C η) * P).eraseLead.SignVariations + 1 := by
+        rw [signvar_eq_eraseLead_add_ite hm, ← hsηq0, ← h_sq0_sηq0, h₂]
+        simp [← hsηq1, hηq1, h_sq0_pos]
+
+      have hLQ2 : ((X - C η) * P).eraseLead.SignVariations =
+          ((X - C η) * P.eraseLead).SignVariations := by
+        apply signvar_ereaseLead_mul_XC_eq_XC_mul_eraseLead hη hqpos
+        exact sign_eq_neg_one_iff.mp (SignType.neg_eq_neg_one ▸ hcsq1 ▸ hsq1)
+
+      linarith
+
+    · --sq1 is positive
+      -- P starts with [+,+,...]. (X-C)*P starts with [+,?,...].
+      -- After dropping the lead of P, this becomes [+,...] and [+,...].
+      -- So the sign variations on P are unchanged when we induct, while (X-C)*P can only lose at
+      -- most one sign change.
+      have hLQ : ((X - C η) * P.eraseLead).SignVariations ≤ ((X - C η) * P).SignVariations := by
+        apply signvar_mul_eraseLead_le_of_nextCoeff hqpos
+        simp [← hsq1, hcsq1, ← sign_eq_one_iff]
+
+      linarith
+
+variable {α : Type*} [LinearOrderedCommRing α]
+
+/-- Descartes' Rule of Signs: the number of positive roots is at most the number of sign
+variations. -/
+theorem descartes_rule_of_signs (P : Polynomial α) : P.roots.countP (0 < ·) ≤ SignVariations P := by
```

Comment:
I think the maintainers generally prefer theorems to be named with the naming conventions, (e.g. `roots_countP_neg_le_SignVariations`) even when the theorem is named.